### PR TITLE
DD4hep: SimTracker/TrackerMaterialAnalysis SpecPars Groups

### PR DESCRIPTION
**Pull request by vargasa**
#### PR description:

Custom made groups added to PhaseI and Phase2 scenarios as discussed in previous Geometry meeting.

FYI: @ianna, @kpedro88, @cvuosalo, @bsunanda 

#### Validation:

```
python Configuration/Geometry/scripts/generate2021Geometry.py -t
python Configuration/Geometry/scripts/generate2026Geometry.py -t
```

Related: https://hypernews.cern.ch/HyperNews/CMS/get/geometry/354.html

### CODE REVIEW

File: Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D49.xml [hunk 1]
```diff
@@ -38,6 +38,7 @@
     <Include ref='Geometry/TrackerSimData/data/PhaseII/TiltedTracker404/trackersens.xml'/>
     <Include ref='Geometry/TrackerSimData/data/PhaseII/TiltedTracker404/pixelsens.xml'/>
     <Include ref='Geometry/TrackerRecoData/data/PhaseII/TiltedTracker613_MB_2019_04/trackerRecoMaterial.xml'/>
+    <Include ref='SimTracker/TrackerMaterialAnalysis/data/trackingMaterialGroups_ForPhaseII.xml'/>
```
Comment (mmusich) [inline Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D49.xml:41]: why is this done only for geometry D49?
There are several active geometries in phase-2. At the bare minimum I would support also the geometries which feature the new Tracker geometries T21, T22 and T23, which are D63,D64 and D65

**Comment by mmusich** [comment]
@vargasa I am slightly at loss with the purpose of this PR.
I think ideally we would like to keep the same functionality as:
```python
# Add our custom detector grouping to DDD
process.XMLIdealGeometryESSource.geomXMLFiles.extend(['SimTracker/TrackerMaterialAnalysis/data/trackingMaterialGroups_ForPhaseII.xml'])
```

also in DD4HEP. Can you clarified what has been decided? 

@mtosi

**Comment by vargasa** [comment]
@mmusich, this is done in order to be able to use `SimTracker/TrackerMaterialAnalysis` with `DDDetectorESProducer`, functionality you pointed out is not currently available for DD4hep, probably @cvuosalo, @ianna can expand on that. Current changes are done as suggested by @kpedro88, @bsunanda, see hn link and notes from previous Simulation meeting

File: Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D49.xml (hunk 1 above)
Comment (vargasa) [reply Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D49.xml:41]: That can be done as well, let's see if we can agree on a solution before extending changes to further geometries. I agreed on submitting a PR to see how the changes "would look like"

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please test
(to see if unit test will pass after #31082 was merged)

**Comment by kpedro88** [comment]
@vargasa please make sure you run the geometry script whenever you modify the geometry dictionary contents. This generates the geometry python/xml files in a consistent way.

**Comment by vargasa** [comment]
Looks like the test is requiring D50, D53, and D60 to be modified as well

```
===== Test "test2026Geometry" ====
cmsExtendedGeometry2026D50XML_cfi.py differs
cmsExtendedGeometry2026D50.xml differs
cmsExtendedGeometry2026D53XML_cfi.py differs
cmsExtendedGeometry2026D53.xml differs
cmsExtendedGeometry2026D60XML_cfi.py differs
cmsExtendedGeometry2026D60.xml differs

---> test test2026Geometry had ERRORS
```

**Comment by vargasa** [comment]
`python Configuration/Geometry/scripts/generate2026Geometry.py -t` is not complaining now locally

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by vargasa** [comment]
@mmusich it looks like `.extend` was only being used by `TrackerMaterialAnalysis` so as long as the changes make it work, it should be fine (IMO)
Screenshot_2020-08-14  CMSSW_11_2_X_2020-08-13-2300 general search geomXMLFiles extend
@kpedro88 , @cvuosalo Is this good to go?

**Comment by kpedro88** [comment]
+upgrade
including additional xml file for material study shows no apparent impact on regular workflows

**Comment by cvuosalo** [comment]
+1

**Comment by mmusich** [comment]
@vargasa, I beg to differ. 
Just because it was the unique caller that doesn't mean that it was not a useful feature. I would like someone from the geometry group to elaborate why this feature is not desirable and / or maintainable in DD4HEP. 
Your solution is probably working (by the way do you have proof that's working correctly?), but it's essentially a hack (IMO). Also if we go this way, every time a new phase-2 geometry is created the SpecPars groups should be added as well to list of XMLs. I am adding @ghugo83 for her information.

**Comment by civanch** [comment]
+1

**Comment by qliphy** [comment]
+1

File: Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml [hunk 2]
```diff
@@ -38,6 +38,7 @@
     <Include ref='Geometry/TrackerSimData/data/PhaseII/TiltedTracker404/trackersens.xml'/>
     <Include ref='Geometry/TrackerSimData/data/PhaseII/TiltedTracker700/pixelsens.xml'/>
     <Include ref='Geometry/TrackerRecoData/data/PhaseII/TiltedTracker700/trackerRecoMaterial.xml'/>
+    <Include ref='SimTracker/TrackerMaterialAnalysis/data/trackingMaterialGroups_ForPhaseII.xml'/>
```
Comment (mmusich) [inline Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml:41]: @vargasa I do not think this update has been appropriate.
Checking e.g. in IB relvals wf. 30234.0 which features the T23 geometry (3D pixels) I see a lot of warnings of the type:
```
Could not process q-name of a DDLogicalPart, reason:
No regex-match for namespace=pixel  name=BModule1Layer1InnerPixelActive

SpecPar selection is:
//pixbar:Phase2PixelBarrel/pixel:Layer1/pixel:RodUnflipped1/pixel:BModule1Layer1/pixel:BModule1Layer1InnerPixelwafer/pixel:BModule1Layer1InnerPixelActive

%MSG
%MSG-e Specific:  (NoModuleName) 19-Aug-2020 07:56:15 CEST pre-events
Could not process q-name of a DDLogicalPart, reason:
No regex-match for namespace=pixel  name=BModule2Layer1InnerPixelActive

SpecPar selection is:
//pixbar:Phase2PixelBarrel/pixel:Layer1/pixel:RodUnflipped1/pixel:BModule2Layer1/pixel:BModule2Layer1InnerPixelwafer/pixel:BModule2Layer1InnerPixelActive

%MSG
%MSG-e Specific:  (NoModuleName) 19-Aug-2020 07:56:15 CEST pre-events
Could not process q-name of a DDLogicalPart, reason:
No regex-match for namespace=pixel  name=BModule3Layer1InnerPixelActive
```

which are present in [CMSSW_11_2_X_2020-08-18-2300](https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_amd64_gcc820/CMSSW_11_2_X_2020-08-18-2300/pyRelValMatrixLogs/run/30234.0_TTbar_14TeV+2026D65+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal/step3_TTbar_14TeV+2026D65+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log), bug e.g. not in [CMSSW_11_2_X_2020-08-16-0000](https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_amd64_gcc820/CMSSW_11_2_X_2020-08-16-0000/pyRelValMatrixLogs/run/30234.0_TTbar_14TeV+2026D65+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal/step3_TTbar_14TeV+2026D65+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log) which does not contain this PR.
Please either clean or fix.

Comment (vargasa) [reply Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml:41]: Hi @mmusich, I added D64, D63, and D65 on your request https://github.com/cms-sw/cmssw/pull/31112#discussion_r468531119 and it seems that wf was not verified by the bot. Do you see any problem with the D64 and D63? If so I can revert https://github.com/cms-sw/cmssw/pull/31112/commits/336ef8d1852d2edd160ce46a30971977b0f4c685 if not I'll revert changes on D65 only :+1:

Comment (kpedro88) [reply Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml:41]: @vargasa the PR tests run fewer WFs than the IB tests, but additional WFs can be added to PR tests on demand.

Comment (mmusich) [reply Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml:41]: @vargasa, the others look OK. For the moment you can just revert D65, but if we go this way we should find a way to fix it....

**Pull request merged**